File: Storage/Code/35.搜索插入位置.py
#
# @lc app=leetcode.cn id=35 lang=python3
#
# [35] 搜索插入位置
#

# @lc code=start
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        l=len(nums)
        left=0
        right=l-1
        if target>nums[right]:
            return l
        while left<right:
            # 向下取整
            mid=left+(right-left)//2
            if nums[mid]==target:
                return mid
            elif nums[mid]<target:
                left=mid+1
            else:
                right=mid
        return left
# @lc code=end

File: Storage/Code/35.搜索插入位置.cpp
/*
 * @lc app=leetcode.cn id=35 lang=cpp
 *
 * [35] 搜索插入位置
 */

// @lc code=start
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        // 二分模板
        // 找第一个大于等于target的位置
        int left=0, right=nums.size()-1;
        // 特殊判断
        if(target>nums[right])  return right+1;
        while(left<right){
            int mid=left+((right-left)>>1);
            if(nums[mid]==target){
                return mid;
            }else if(nums[mid]>target){
                right=mid;
            }else{
                left=mid+1;
            }
        }
        return left;
    }
};
// @lc code=end

